File: 2022/R/02.R
#!/usr/bin/env -S Rscript --vanilla

# info --------------------------------------------------------------------

#     win  = 6 | r > s; s > p; p > r
#     tie  = 3 | r == r; s == s; r == r
#     loss = 0 | s < r; p < s; r < p
#     rock = 1 | A X
#    paper = 2 | B Y
# scissors = 3 | C Z

# A Y = 6 + 2 = 8
# B X = 0 + 1 = 1
# C Z = 3 + 3 = 6
# total: 15

# x = lose
# y = draw
# z = win
#
# A Y = 1r + 3t = 4
# B X = 1r + 0l = 1
# C Z = 1r + 6w = 7
# total: 12


# solve -------------------------------------------------------------------

x <- read.table("2022/data/02", col.names = c("opponent", "me"))

mat <- function(..., by_row = FALSE) {
  matrix(
    as.integer(unlist(c(...))),
    byrow = by_row,
    ncol = 3L,
    dimnames = list(
      opponent = c("A", "B", "C"),
      me = c("X", "Y", "Z")
    )
  )
}

count <- with(x, table(opponent, me))
result <- mat(3L, 0L, 6L, 6L, 3L, 0L, 0L, 6L, 3L)
play <- mat(rep(1:3, each = 3))

solution1 <- sum(count * (result + play))

result2 <- mat(rep(0:2 * 3, each = 3))
play2 <- mat(3L, 1L, 2L, 1L, 2L, 3L, 2L, 3L, 1L)

solution2 <- sum(count * (result2 + play2))


# test --------------------------------------------------------------------

solutions <- readLines("2022/solutions/02")

stopifnot(
  solution1 == solutions[1],
  solution2 == solutions[2]
)

File: 2022/python/02.py
#! /usr/bin/env python3

import pandas as pd
import numpy as np


def main() :

    df = pd.read_table("2022/data/02", header=None, sep=" ", names=["opponent", "me"])

    counts = df.\
        value_counts().\
        sort_index().\
        to_frame().\
        pivot_table(
            index="me",
            columns="opponent",
            values=0,
            fill_value=0
        )

    counts = np.asarray(counts)
    result = np.array([[3, 0, 6], [6, 3, 0], [0, 6, 3]], dtype=int)
    play = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]], dtype=int)

    solution1 = np.sum(counts * [result + play])

    result2 = np.array([[0, 0, 0], [3, 3, 3], [6, 6, 6]], dtype=int)
    play2 = np.array([[3, 1, 2], [1, 2, 3], [2, 3, 1]], dtype=int)

    solution2 = np.sum(counts * (result2 + play2))

    return [solution1, solution2]


if __name__ == "__main__" :
    results = main()
    solutions = np.loadtxt("2022/solutions/02")

    if any(results != solutions):
        raise ValueError("solutions are incorrect")
